src/NHibernate/Proxy/DynamicProxy/ProxyFactory.cs

```diff
@@ -35,62 +37,43 @@ public ProxyFactory(IProxyAssemblyBuilder proxyAssemblyBuilder)
 			: this(new DefaultyProxyMethodBuilder(), proxyAssemblyBuilder) {}
 
 		public ProxyFactory(IProxyMethodBuilder proxyMethodBuilder)
-			: this(new DefaultyProxyMethodBuilder(), new DefaultProxyAssemblyBuilder()) {}
+			: this(proxyMethodBuilder, new DefaultProxyAssemblyBuilder()) {}
 
 		public ProxyFactory(IProxyMethodBuilder proxyMethodBuilder, IProxyAssemblyBuilder proxyAssemblyBuilder)
 		{
-			if (proxyMethodBuilder == null)
-			{
-				throw new ArgumentNullException("proxyMethodBuilder");
-			}
-			ProxyMethodBuilder = proxyMethodBuilder;
+			ProxyMethodBuilder = proxyMethodBuilder ?? throw new ArgumentNullException(nameof(proxyMethodBuilder));
 			ProxyAssemblyBuilder = proxyAssemblyBuilder;
-			Cache = new ProxyCache();
 		}
 
-		public IProxyCache Cache { get; private set; }
+		//Since v5.1
+		[Obsolete("This property is not used anymore and will be removed in a next major version")]
+		public IProxyCache Cache { get; } = new ProxyCache();
 
-		public IProxyMethodBuilder ProxyMethodBuilder { get; private set; }
+		public IProxyMethodBuilder ProxyMethodBuilder { get; }
 
-		public IProxyAssemblyBuilder ProxyAssemblyBuilder { get; private set; }
+		public IProxyAssemblyBuilder ProxyAssemblyBuilder { get; }
 
 		public object CreateProxy(System.Type instanceType, IInterceptor interceptor, params System.Type[] baseInterfaces)
 		{
 			System.Type proxyType = CreateProxyType(instanceType, baseInterfaces);
 			object result = Activator.CreateInstance(proxyType);
 			var proxy = (IProxy) result;
 			proxy.Interceptor = interceptor;
-
 			return result;
 		}
 
 		public System.Type CreateProxyType(System.Type baseType, params System.Type[] interfaces)
 		{
-			System.Type[] baseInterfaces = ReferenceEquals(null, interfaces) ? new System.Type[0] : interfaces.Where(t => t != null).ToArray();
-			
-			TypeInfo proxyTypeInfo;
+			if (baseType == null) throw new ArgumentNullException(nameof(baseType));
 
-			// Reuse the previous results, if possible, Fast path without locking.
-			if (Cache.TryGetProxyType(baseType, baseInterfaces, out proxyTypeInfo))
-				return proxyTypeInfo;
+			var typeFactory = _cache.GetOrAdd(
+				new ProxyCacheEntry(baseType, interfaces),
+				k => CreateUncachedProxyType(k.BaseType, k.Interfaces));
```

Comment:
We have overlooked a potential issue here. `ConcurrentDictionary` does not guarantee the factory method will be called only once per key. It may be called multiple times if concurrent threads attempts to add the same key. `ConcurrentDictionary` only guarantee it will yield the same value to all threads.

So concurrency may leads to try creating the proxy multiple times.
I think we should fix it with this [SO answer](https://stackoverflow.com/a/31637510/1178314).

Reply:
Yes I know about that issue and a solution. No, I did not overlook it. I have considered concurrent dictionary of lazy, end even coded it that way somewhere in between. Unfortunately lazy wasn’t so good for performance reasons. Therefore I decided, that lazy isn’t really necessary there.

Reply:
Do you mean creating multiple time the same proxy and its assembly is safe? I do not know it surely when reading the code and the documentation of assembly building methods.

Reply:
Yes, it is safe. For each proxy we create a new assembly (I think, it is not very efficient, but I’m not sure if we can create only one assembly) with a songle type. There will be no harm if we create the same twice. As I understand, at the end we will end up with a single type that were created faster and filled the place is a dictionary.